### ISSUE TITLE
Mounting a nonexistent host path, even as readonly, creates a directory on the host

### ISSUE THREAD

**Issue by rohansingh** [docker/docker#13121]
This behavior seems to have been around for a while but is unexpected. Try mounting a nonexistent host path as a volume:

    docker run -v /abc:/mnt busybox

The `/abc` directory will be created on the host:

```
docker@dev:/$ ls
abc/     dev/     home/    lib/     linuxrc  opt/     root/    sbin/    tmp      var/
bin/     etc/     init     lib64    mnt/     proc/    run/     sys/     usr/
```

This happens even if you attempted to mount the host path as readonly (`-v /abc:/mnt:ro`), which is even more unexpected. This behavior is particularly problematic when attempting to mount a file which may or may not exist — Docker creates a directory at the file path, which can cause problems down the road.

This behavior seems to be quite old (since at least 1.2.0), but doesn't seem to be documented.

**Comment by thaJeztah** [comment]
I think it would make sense to "bail out" if the mount-path doesn't exist. This would of course, be a breaking change, so will have to be decided on.

/cc @crosbymichael @cpuguy83 @calavera

**Comment by rohansingh** [comment]
Yeah, definitely would be a change in behavior. I'd argue that at least in the read-only case, Docker should never modify the host fs.

**Comment by thaJeztah** [comment]
I think the discussion there would be if `:ro` applies to anything **inside** the mount, or to the mount itself as well :) 

But in general, I think that producing an error is the right approach; In addition to creating directories on the host (out of control of the user), I've been bitten a few times as well when I had a typo in my host-path and the container just seemed to start okay (with an empty volume)

**Comment by vdemeester** [comment]
@thaJeztah I think there should be at least a warning if not an error if the path on the host does not exist. It's a behaviour that feels a bit strange for users (I tend to have the question more than once in training session :stuck_out_tongue_winking_eye:).

I might take a look and make a proposition PR on this one if this is alright with others

**Comment by thaJeztah** [comment]
@vdemeester :+1: please search for existing issues as well; I think there have been some in the past (but they may be closed).

**Comment by yourteamneedsadam** [comment]
Plus one for this being confusing. A couple days ago I rebooted my workstation to install updates and some directories magically appeared in my home folder. After a lot of guessing I eventually figured out docker was creating them when it started. I had stopped containers who had originally been run with those local paths mounted (with `-v`). That was hard enough to diagnose on a minimal system, on heavy systems it could be a lot worse. Related to (or possibly a duplicate of) #13468.

I originally found this on XUbuntu 14.04 Desktop but reproduced it on OSX running boot2docker. To reproduce:

1. Create a temp directory.
2. Run a container with `-v /path/to/temp:/mnt`
3. Stop the container.
4. Run `docker ps` and double-check that it's stopped.
5. Delete the temp directory.
6. Restart the docker daemon (or boot2docker).
7. The directory comes back.
8. If you repeat those steps but `docker rm` the container before restarting docker, the directory doesn't come back.

**Comment by cpuguy83** [comment]
This behavior is detailed in: https://docs.docker.com/userguide/dockervolumes/
In most cases it is expected behavior and not something we can really change w/o breaking lots of people.
When mounting as `ro`, this is implied as `ro` in the container, not to docker.

Although I do agree that this should have been a case where there is an error thrown if the hostpath doesn't exist, unfortunately this one got through.

I'd recommend in most cases not doing host binds like this, especially with the enhanced volume management features available today (named volumes in the current release, `docker volume` subcommand in 1.9).

**Comment by thaJeztah** [comment]
Closing, because https://github.com/docker/docker/pull/16349 was merged, and this behavior is scheduled to be removed in docker 1.11

**Comment by biggianteye** [comment]
Looks like it was reverted, which is a shame: https://github.com/docker/docker/pull/16349#issuecomment-231603065

**Comment by cpuguy83** [comment]
@biggianteye Indeed.
We've added a new API for mounts which will never create a host path. Still waiting on getting the CLI options to actually use the new API.

**Comment by biggianteye** [comment]
@cpuguy83 Ah, that's great to hear. Could you point me to some documentation for the new API so I can read up on it?

**Comment by cpuguy83** [comment]
It's in the generated swagger docs, but the yaml is probably easier to read since `/containers/create` is so big.
https://github.com/docker/docker/blob/v1.13.1/api/swagger.yaml#L220

**Comment by dragon788** [comment]
**TL;DR**
Use `--mount` and not `-v` if using bind mounts, or better yet use real volumes with `docker volume create --name my-data-volume`.

For anybody else running across this from Google using the first example below will emit a warning and stop while the second example will create a folder and continue causing painful issues later.

```
docker run -d \
  -it --rm \
  --name devtest \
  --mount type=bind,source="$(pwd)"/myfakedir,target=/app \
  busybox
```

This creates the 'myfakedir' in the current directory owned by root. 

```
docker run -d \
  -it --rm \
  --name devtest \
  -v "$(pwd)"/myfakedir:/app \
  busybox
```

**Comment by h45rd** [comment]
Is it expected behavior that when using the _--mount_ flag the existence of the source directory is only checked on creation of the container, not on execution?

If so, might I propose to change that behavior?
It seems inconsistent and quite unexpected, at least to me.

I thought creating my containers with the _--mount_ flag would prevent them from starting if the specified source directory was absent (e.g. when a network mount became unavailable).
Instead they behave in the same way as they would when using the _-v_ flag and create the specified source directory tree on execution.

**Comment by thaJeztah** [comment]
@h45rd Thanks for reporting that; I don't think that _should be_ the intended behaviour.

I'm able to reproduce using the following steps;

```bash
$ mkdir repro-13121 && cd repro-13121
$ docker create --name repro-13121 \
  --mount type=bind,source="$(pwd)"/doesnotexist,target=/app \
  busybox
Error response from daemon: invalid mount config for type "bind": bind source path does not exist
```

During create, an error is produced (as expected)

First creating the directory makes the `docker create` pass;

```bash
$ mkdir doesnotexist
$ docker create --name repro-13121 \
   --mount type=bind,source="$(pwd)"/doesnotexist,target=/app \
   busybox
a7b4d510d1afd3eabea59414b94e9565af2927606f8a79f27aa855f3a8ea2817
```

Now, remove the directory, and verify the directory is gone;

```bash
$ rm -r doesnotexist
$ ls
(empty)
```

Starting the container creates the host directory;

```bash
$ docker start repro-13121
repro-13121
$ ls
doesnotexist
```

I'll create an issue for that

**Comment by thaJeztah** [comment]
Opened https://github.com/moby/moby/issues/35821 for tracking

**Comment by pstrzelczak** [comment]
Why is this ticket closed?
```
docker run -v /abc:/mnt busybox
```
still silently creates /abc directory with root privileges (checked on 17.05.0-ce and 17.06.2.ce) is such path did not exist before.

**Comment by thaJeztah** [comment]
@pstrzelczak that's the expected behaviour when using the `-v <host-path>:<container-path>` notation. See https://github.com/moby/moby/pull/21666

**Comment by gun1x** [comment]
This is really annoying when using ansible to create containers... is there any way to avoid this behavior?

**Comment by cpuguy83** [comment]
@gun1x the solution is mentioned above.
You need to use the "-mount" syntax.

**Comment by teresaejunior** [comment]
@gun1x For Ansible: https://github.com/ansible/ansible/pull/49808

**Comment by weliveindetail** [comment]
> @biggianteye We've added a new API for mounts which will never create a host path. Still waiting on getting the CLI options to actually use the new API.

Actually, the old behavior was quite useful to mount container paths back to the host! Is there any way to do it with to new API? My use-case was cross-compilation:
https://weliveindetail.github.io/blog/post/2021/05/19/remote-cross-jit-mongoose.html#cross-compile-bitcode-for-alpine-linux

Thanks

**Comment by thaJeztah** [comment]
>  Is there any way to do it with the new API

@weliveindetail Yes, in the v23.0.0 release (API v1.42) we added an option to create the host-path with the new API (but it's not the _default_, and not currently exposed in the CLI);

- https://github.com/moby/moby/pull/43484
- https://github.com/moby/moby/issues/43483

> My use-case was cross-compilation:

If you use it to _build_ things, you may also be interested in using the `--output` option that's provided by `docker build` with BuildKit enabled (which is now the default for Docker 23.0.0, which uses `buildx` as a default). With that option, you can use a `docker build` that builds the (multi-stage) Dockerfile, but instead of _tagging_ an image, copies the files to the host after completion. It currently requires you to produce a stage that _only_ contains the binaries you need (so usually would be a `FROM scratch` stage), but there are some feature requests to improve that (I think https://github.com/moby/buildkit/issues/874 is tracking this, but there may be others).

For example:


```dockerfile
# syntax=docker/dockerfile:1

FROM alpine AS build
WORKDIR /myproject/release
RUN echo "I am some build-articact" > binary1.txt
RUN echo "I am another build-articact" > binary2.txt

# this stage only contains the binaries built, it's intended for
# exporting (using --output).
FROM scratch AS release-binaries
COPY --from=build /myproject/release/ .
```

Then build the Dockerfile, and use the `--output` option to specify where the files should be copied to;

```bash
docker build --output=./release/ .

# [+] Building 4.6s (13/13) FINISHED
# => [internal] load build definition from Dockerfile                   0.0s
# => => transferring dockerfile: 394B                                   0.0s
# ....
# => [release-binaries 1/1] COPY --from=build /myproject/release/ .     0.0s
# => exporting to client                                                0.0s
# => => copying files 123B                                              
```

After the build finishes, the files can be found in that location:

```bash
tree
.
├── Dockerfile
└── release
    ├── binary1.txt
    └── binary2.txt

1 directory, 3 files
```

Using this approach allows you to build everything in a sandboxed environment (the `Dockerfile` steps used in the `docker build` do not have direct access to your local paths, and use a temporary copy of the files, so more reproducible), and unlike the `docker run` (with `--mount`, or `-v`), you can also use this with a _remote_ builder (the daemon can run elsewhere, and the files get copied back to your local machine). This also allows using a builder that's natively running on a different architecture (if you don't want to cross-compile).

Also see (those examples use `docker buildx build`, but on Docker 23.0 the default is for `docker build` to be an alias for `docker buildx build`);

- https://docs.docker.com/build/exporters/local-tar/
- https://docs.docker.com/engine/reference/commandline/buildx_build/#output

**Comment by weliveindetail** [comment]
@thaJeztah Thanks for your quick response! Interesting, it's a very different approach, but it looks like that should do it. Will give it a try soon!
